File: cloud-storage-spring-java/src/main/java/com/github/sevtech/cloud/storage/spring/service/AbstractStorageService.java
package com.github.sevtech.cloud.storage.spring.service;

import com.github.sevtech.cloud.storage.spring.exception.NoBucketException;
import com.github.sevtech.cloud.storage.spring.model.UploadFileRequest;
import lombok.extern.slf4j.Slf4j;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.util.Optional;

@Slf4j
public abstract class AbstractStorageService {

    protected InputStream clone(final InputStream inputStream) {
        InputStream result = null;
        try {
            inputStream.mark(0);
            ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
            byte[] buffer = new byte[1024];
            int readLength;
            while ((readLength = inputStream.read(buffer)) != -1) {
                outputStream.write(buffer, 0, readLength);
            }
            inputStream.reset();
            outputStream.flush();
            result = new ByteArrayInputStream(outputStream.toByteArray());
        } catch (final Exception exception) {
            log.warn("Exception cloning inputStream", exception);
        }
        return result;
    }

    protected String getBucketName(final String bucketName, final String defaultBucketName) throws NoBucketException {
        return Optional.ofNullable(
                Optional.ofNullable(bucketName).orElse(defaultBucketName))
                .orElseThrow(() -> new NoBucketException("Bucket name not indicated"));
    }

    protected String getFilePath(final UploadFileRequest request) {
        return request.getFolder().concat("/").concat(request.getName());
    }

}

File: cloud-storage-spring-kotlin/src/main/kotlin/com/github/sevtech/cloud/storage/spring/service/AbstractStorageService.kt
package com.github.sevtech.cloud.storage.spring.service

import com.github.sevtech.cloud.storage.spring.model.UploadFileRequest
import com.github.sevtech.cloud.storage.spring.exception.NoBucketException
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.InputStream
import java.util.*

abstract class AbstractStorageService {

    protected fun InputStream.clone(): InputStream? {
        var result: InputStream? = null
        try {
            this.mark(0)
            val outputStream = ByteArrayOutputStream()
            val buffer = ByteArray(1024)
            var readLength = this.read(buffer)
            while ((readLength) != -1) {
                outputStream.write(buffer, 0, readLength)
                readLength = this.read(buffer)
            }
            this.reset()
            outputStream.flush()
            result = ByteArrayInputStream(outputStream.toByteArray())
        } catch (ex: Exception) {
            ex.printStackTrace()
        }
        return result
    }

    @Throws(NoBucketException::class)
    protected fun getBucketName(bucketName: String?, defaultBucketName: String): String {
        return Optional.ofNullable(
                Optional.ofNullable(bucketName).orElse(defaultBucketName))
                .orElseThrow { NoBucketException("Bucket name not indicated") }
    }

    protected fun getFilePath(request: UploadFileRequest): String {
        return request.folder + "/" + request.name
    }
}